### ISSUE TITLE
[QUESTION] How to display an error with the `max` constraint?

### ISSUE THREAD

**Issue by opack** [ciscoheat/sveltekit-superforms#283]
Hello!
Zod has a `max` constraint that can make sure the string does not have more than the expected number or characters.
When using SuperForms, this constraint is turned into a `maxlength` attribute.

The problem is that the field just refuses to let you enter more characters, which is not SuperForms' fault, obviously. But using this attribute does not give the user any information as to why he cannot input more characters, because no error message given to the Zod schema is displayed.

Is there any way to workaround this situation and make sure the user can input a wrong value, obtain a helpful message and correct his input (as with the other constraints)?

**Comment by ciscoheat** [comment]
Hi, unless it works with the [customValidity option](https://superforms.rocks/concepts/error-handling#customvalidity) on the field, I think you must disable the constraint and set a custom error message with [setCustomValidity](https://developer.mozilla.org/en-US/docs/Web/API/HTMLObjectElement/setCustomValidity). The `$errors` store could be helpful there, when displaying a friendly error message.

**Comment by opack** [comment]
Thanks for the quick answer! The `customValidity` option does the trick indeed of not setting the `maxlength`, but the problem is that I would need to wire the Zod checks by myself, which would clutter my code. I love using Superforms precisely to make all boilerplaty non-business code disappear 😅
I'll see what I can do! Thanks for pointing me in the right direction, and thanks again for bringing us SuperForms 👏

**Comment by ciscoheat** [comment]
Thank you for your kind words. :) Do you really need to handle the Zod checks yourself? If you're using the [validators](https://superforms.rocks/concepts/client-validation#validators) option, it should be pretty much automatic.

**Comment by opack** [comment]
After reading your first answer, I thought that I would need to call the `validate` method of the Zod schema for each field myself, and use the result on my own to display it. But I was wrong (hopefully!).

As you suggested after, I just have to keep the `validators` option (that I was already using) and remove the `{...$constraints}` that I applied to the `<input>`. By doing that, my fields are correctly validated, the error messages do appear correctly and the user can type what he wants in the field and receive some feedback as to what is wrong 👍 This is exactly what I wanted, thanks for posting your second message 🙏 

The only difference is that the `<input>` no longer uses the HTML attributes. I can live with that. because on top of my initial problem I gain something great: the possibility to have multiple regexps!

That was initially an issue for me, as I wanted the user to have correct feedback to guide him during the input. Just saying "your password is incorrect because it does not as the correct format" is not very friendly. I prefered being able to tell him "Missing a digit, missing a special character...".
I initially browsed the issues here and saw that the topic had already been discussed. I then settled with only one regexp in my Zod schema, because anyway only one would appear in the `pattern` attribute.
But now that I don't use this attribute anymore, I can have multiple regexps (and bonus: they are really short and simple), each one with its own error message (just tested it and it's great!).

So, once again, thank you very much 😁
